packages/treat/src/webpack-plugin/treatCompiler.js

```diff
@@ -106,48 +99,86 @@ function getRootCompilation(loader) {
-        new NodeTargetPlugin(),
-        new SingleEntryPlugin(loader.context, loader.resourcePath),
-        new LimitChunkCountPlugin({ maxChunks: 1 }),
-        new ExternalsPlugin('commonjs', 'treat'),
-      ],
     );
 
-    const subCache = 'subcache ' + __dirname + ' ' + loader.resourcePath;
-    childCompiler.hooks.compilation.tap(TWL, (compilation) => {
-      if (compilation.cache) {
-        if (!compilation.cache[subCache]) compilation.cache[subCache] = {};
-        compilation.cache = compilation.cache[subCache];
-      }
-    });
+    const NodeTemplatePlugin = compat.getNodeTemplatePlugin(loader._compiler);
+    const NodeTargetPlugin = compat.getNodeTargetPlugin(loader._compiler);
+    const LimitChunkCountPlugin = compat.getLimitChunkCountPlugin(
+      loader._compiler,
+    );
+    const ExternalsPlugin = compat.getExternalsPlugin(loader._compiler);
+
+    new NodeTemplatePlugin(outputOptions).apply(childCompiler);
+    new NodeTargetPlugin().apply(childCompiler);
+
+    if (compat.isWebpack5) {
+      const {
+        EntryOptionPlugin,
+        library: { EnableLibraryPlugin },
+      } = loader._compiler.webpack;
+
+      new EnableLibraryPlugin('commonjs2').apply(childCompiler);
+
+      EntryOptionPlugin.applyEntryOption(childCompiler, loader.context, {
+        child: {
+          library: {
+            type: 'commonjs2',
+          },
+          import: [loader.resourcePath],
+        },
+      });
+    } else {
+      // Webpack 4 code. Remove once support is removed
+      const { LibraryTemplatePlugin, SingleEntryPlugin } = require('webpack');
+
+      new LibraryTemplatePlugin(null, 'commonjs2').apply(childCompiler);
+      new SingleEntryPlugin(loader.context, loader.resourcePath).apply(
+        childCompiler,
+      );
+    }
 
-    let source;
+    new LimitChunkCountPlugin({ maxChunks: 1 }).apply(childCompiler);
+    new ExternalsPlugin('commonjs', 'treat').apply(childCompiler);
 
-    childCompiler.hooks.afterCompile.tapAsync(TWL, (compilation, callback) => {
-      source =
-        compilation.assets[loader.resourcePath] &&
-        compilation.assets[loader.resourcePath].source();
+    let source;
 
-      // Remove all chunk assets
-      compilation.chunks.forEach((chunk) => {
-        chunk.files.forEach((file) => {
-          delete compilation.assets[file];
+    if (compat.isWebpack5) {
+      childCompiler.hooks.compilation.tap(TWL, (compilation) => {
+        compilation.hooks.processAssets.tap(TWL, () => {
+          source =
+            compilation.assets[loader.resourcePath] &&
+            compilation.assets[loader.resourcePath].source();
+
+          // Remove all chunk assets
+          compilation.chunks.forEach((chunk) => {
+            chunk.files.forEach((file) => {
+              compilation.deleteAsset(file);
+            });
+          });
         });
       });
-
-      callback();
-    });
+    } else {
+      childCompiler.hooks.afterCompile.tap(TWL, (compilation) => {
```

Comment:
This is no longer async?

Reply:
Doesn't need to be. Probably a mistake in the first place.